Resolve #552: events stick under certain conditions

ashes999 opened the pull request:
Note: this only works with the following HTML mark-up:

`<body onblur='Crafty.resetKeyDown();'>`

This is despite the DOM event bindings for `blur`. Please provide feedback on how to fix this (it should work, but doesn't).

Commit:
Resolved #552
- Entities stop moving when focus is lost (eg. `CTRL+F`, `CTRL+E`)
- Blur correctly traps all keys, except `CTRL+H`
- Various fixes from code-review
- Tested in IE, Chrome, and FF

starwed commented on src/controls.js:
```diff
@@ -27,18 +27,39 @@ Crafty.extend({
+                
+        // Trigger direction change event
+        this.trigger('NewDirection', this._movement);   
+    },
```
Looks like there's some leftover code from the multiway fix here.  That should all be taken care of by triggering the keyup events.

ashes999 replied:
You're right. I'll integrate this change.

starwed commented on src/controls.js:
```diff
@@ -27,18 +27,39 @@ Crafty.extend({
+                 });
+             }
+        }
+        Crafty.keydown = [];
```
Use `Crafty.keydown.length=0` to quickly wipe an array without creating a new one.

ashes999 replied:
It's interesting. Changing this to `.keydown.length = 0` creates a side-effect: the first case of `up+ctrl+e` halts correctly, but it remembers the state of the up key being held down, and subsequent calls still make the fourway stick.

I'll have to leave it as-is for now.

starwed commented:
> This is despite the DOM event bindings for blur. Please provide feedback on how to fix this (it should work, but doesn't).

You probably want to use `window` rather than `document.body` when binding that event, although I haven't gone in and tested that!  :)

The basic idea here looks good!    Some comments:
- If a user holds down a key while the window loses focus, and then lets up when focus is returned, it'll cause the opposite problem.  So in `keyboardDispatch` we need to not fire a keyup event if Crafty doesn't know about the key being down.  We already do the equivalent check for keydown.
- I think you added some extraneous spaces at some line endings.  If you've installed the grunt toolchain, run `grunt check` and it'll lint the code -- that'll tell you which lines need fixing up.  (And also, run unit tests.  To just run the linter by itself, use `grunt jshint`.)
- If you can, please squash the commits down so that the revision history is cleaner.  ([Instructions here](https://github.com/craftyjs/Crafty/wiki/Rewriting-History) if you're not sure how.)

Thanks for the PR!

ashes999 commented:
I can't get `Crafty.addEvent` to to work; neither `document.body` nor `window` actually bind the event. I'm pretty sure I want to bind to `document.body`, because I want the equivalent of `<body onblur="...">`.

Can you please investigate this? It seems like another bug.

starwed commented:
> I can't get Crafty.addEvent to to work; neither document.body nor window actually bind the event. I'm pretty sure I want to bind to document.body, because I want the equivalent of <body onblur="...">.

I checked out your branch -- changing `document.body` to `window` when adding the blur event worked fine for me in Firefox and Chrome.  Maybe doublecheck that it wasn't working?  For instance, maybe your browser was caching the old version when you tested it.   (As I understand it, the standard is that the window fires a blur event, and that's what we actually care about -- as long as the window is focused, it's catching the keyup/down events.)

> It's interesting. Changing this to .keydown.length = 0 creates a side-effect: the first case of up+ctrl+e halts correctly, but it remembers the state of the up key being held down, and subsequent calls still make the fourway stick.

Ah, that's because the keydown map starts as an object.  If you change it so that it's initalized as an array, then setting the length to 0 should work fine.  :)

starwed commented:
Oh, and the reason adding `onblur` to the body element works can be found [in the HTML5 spec](http://www.w3.org/html/wg/drafts/html/master/sections.html#the-body-element):

> The body element exposes as event handler content attributes a number of the event handlers of the Window object. It also mirrors their event handler IDL attributes.
> 
> The **onblur**, onerror, onfocus, onload, onresize, and onscroll event handlers **of the Window object**, exposed on the body element, replace the generic event handlers with the same names normally supported by HTML elements.

In other words, the onblur _attribute_ of `<body>` attaches the event to the window object.  Using `addEventListener` apparently doesn't work quite the same way.   

> A regular event listener attached to the body using addEventListener(), however, would be run when the event bubbled through the body and not when it reaches the Window object:

(I only found this out through this [stackoverflow question](http://stackoverflow.com/questions/6939047/does-html-body-tag-have-an-onblur-event) )

ashes999 commented:
You're right. It seems to work as expected now.

For the object, since it's an object (not an array), reseting it to a new object (`{ }`) should be sufficient.

Please verify. I think we're done here :)

ashes999 commented:
Just to clarify:
- I updated, squashed, and re-pushed
- There is still an occasional weird issue with focus. But it's not related to this (just like `CTRL+H` still breaks `fourway`)

ashes999 commented:
Any comments @starwed ?

starwed commented:
@ashes999  Nah, I think this looks good.  Just want to merge develop into testing before this lands, since it's a bigger change.

starwed commented:
We still need to fix the CMD key issue on macs, but this patch is nicely compatible with that.  (We can just call the reset function and do whatever else will happen.)

starwed commented:
Oh, and thanks for the PR!   :)

ashes999 commented:
Glad to help :)

Two open issues are still:
- CMD key (Macs)
- CTRL+H (Windows at least)

Since #552 is closed, I suggest opening one (or two) new issues to track these two.

Thanks for the code review!

starwed commented:
I'll just reopen 552.  (Since there's already discussion of the issues there.)
